### PR TITLE
[Fixes #2190] Remove GetSupportedContentTypes method from IOutputFormatt...

### LINKED ISSUE #2190: Remove GetSupportedContentTypes method from IOutputFormatter and move to a separate metadata interface
https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.Core/Formatters/IOutputFormatter.cs#L36

### PR DESCRIPTION
**Pull request by kichalla**
...er and move to a separate metadata interface

@rynowak

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/Description/IApiFormatterMetadataProvider.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNet.Mvc.Description
+{
+    public interface IApiFormatterMetadataProvider
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Description/IApiFormatterMetadataProvider.cs]: Docs. Also `IApiFormatterMetadataProvider` -> `IApiResponseFormatMetadataProvider`

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Description/IApiFormatterMetadataProvider.cs]: Mention here that this is a metadata api for IOutputFormatters to implement

File: test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/ObjectResultTests.cs [hunk 2]
```diff
@@ -309,72 +309,72 @@ public void ObjectResult_Create_CallsContentResult_InitializesValue()
...
+        //    // Assert
+        //    Assert.Equal(mockCountingFormatter.Object, formatter);
+        //    mockCountingFormatter.Verify(v => v.CanWriteResult(context,
+        //                                                       mockCountingSupportedContentType),
+        //                                                       Times.Once());
+        //    var callCount = (acceptHeaderCollection == null ? 0 : acceptHeaderCollection.Count()) + 1;
+        //    mockCountingFormatter.Verify(v => v.CanWriteResult(context,
+        //                                      It.IsNotIn<MediaTypeHeaderValue>(mockCountingSupportedContentType)),
+        //                                      Times.Exactly(callCount));
+        //}
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/ObjectResultTests.cs]: This isn't right :laughing:

File: test/Microsoft.AspNet.Mvc.Core.Test/Formatters/StreamOutputFormatterTest.cs [hunk 3]
```diff
@@ -11,39 +11,6 @@ namespace Microsoft.AspNet.Mvc
...
-                Assert.Null(contentTypes);
-            }
-            else
-            {
-                Assert.Equal(1, contentTypes.Count);
-                Assert.Equal(expected, contentTypes[0].ToString());
-            }
-        }
-
-        [Theory]
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Formatters/StreamOutputFormatterTest.cs]: Do we have tests that verify the result of `CanWriteResult(..., contentType: null)`?  Please check for all of these

**Comment by rynowak** [comment]
:watch:

File: test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/ObjectResultTests.cs [hunk 4]
```diff
@@ -309,72 +309,72 @@ public void ObjectResult_Create_CallsContentResult_InitializesValue()
...
-            Assert.Equal(mockCountingFormatter.Object, formatter);
-            mockCountingFormatter.Verify(v => v.CanWriteResult(context,
-                                                               mockCountingSupportedContentType),
-                                                               Times.Once());
-            var callCount = (acceptHeaderCollection == null ? 0 : acceptHeaderCollection.Count()) + 1;
-            mockCountingFormatter.Verify(v => v.CanWriteResult(context,
-                                              It.IsNotIn<MediaTypeHeaderValue>(mockCountingSupportedContentType)),
-                                              Times.Exactly(callCount));
-        }
+        //[Theory]
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/ObjectResultTests.cs]: why is this commented?

File: src/Microsoft.AspNet.Mvc.Core/Description/IApiFormatterMetadataProvider.cs (hunk 1 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Description/IApiFormatterMetadataProvider.cs]: :+1:  on `IApiResponseFormatMetadataProvider`

Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Core/Description/IApiFormatterMetadataProvider.cs]: Sure

File: test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/ObjectResultTests.cs (hunk 4 above)
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/ObjectResultTests.cs]: Yeah, fixed this test in next iteration.

**Comment by kichalla** [comment]
Updated. /cc: @rynowak @harshgMSFT

File: test/Microsoft.AspNet.Mvc.Core.Test/Formatters/StreamOutputFormatterTest.cs (hunk 3 above)
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.Core.Test/Formatters/StreamOutputFormatterTest.cs]: I have enabled the commented out test which verifies this now.

File: src/Microsoft.AspNet.Mvc.Core/Description/IApiResponseFormatMetadataProvider.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+using System.Collections.Generic;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNet.Mvc.Description
+{
+    /// <summary>
+    /// Provides metadata information about the response format to an <see cref="IApiDescriptionProvider"/>.
+    /// </summary>
+    /// <remarks>
+    /// An<see cref="IOutputFormatter"/> should implement this interface if it wants to expose metadata information
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Description/IApiResponseFormatMetadataProvider.cs]: `if it wants to` -> `to`

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Description/IApiResponseFormatMetadataProvider.cs]: `An<see cref="IOutputFormatter"/>` spacing

File: test/Microsoft.AspNet.Mvc.Core.Test/Formatters/StreamOutputFormatterTest.cs (hunk 3 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Formatters/StreamOutputFormatterTest.cs]: Please add a test for this.

**Comment by rynowak** [comment]
:shipit: after adding back test and making docs tweaks

**Comment by kichalla** [comment]
@rynowak . I have updated the PR with new tests.
@harshgMSFT

File: src/Microsoft.AspNet.Mvc.Core/Formatters/StreamOutputFormatter.cs [hunk 6]
```diff
@@ -15,32 +15,6 @@ namespace Microsoft.AspNet.Mvc
     /// </summary>
     public class StreamOutputFormatter : IOutputFormatter
     {
-        /// <summary>
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/StreamOutputFormatter.cs]: Do we not want api explorer to show data for this ?

**Comment by harshgMSFT** [comment]
:shipit:

**Pull request merged**